src/NHibernate/Hql/Ast/ANTLR/HqlParser.cs

```diff
@@ -212,17 +212,17 @@ public IASTNode NegateNode(IASTNode node)
 			switch (node.Type)
 			{
 				case OR:
-					node.Type = AND;
-					node.Text = "{and}";
-					NegateNode(node.GetChild(0));
-					NegateNode(node.GetChild(1));
-					return node;
+					IASTNode newNode = (IASTNode) TreeAdaptor.Create(AND, "{and}");
```

Comment:
I don't think new node is actually required. We can keep reusing old node. Just need to set negated nodes as children:

```C#
case OR:
	node.Type = AND;
	node.Text = "{and}";
	node.SetChild(0, NegateNode(node.GetChild(0)));
	node.SetChild(1, NegateNode(node.GetChild(1)));
	return node;
```

Reply:
I didn't check the SetChild-Methode... If it replaces, that variant would also be fine.